## Task 3 - Cory Wilson

In [14]:
# Initial Import and Dataframe creation

import pandas as pd
import numpy as np
from mlxtend.preprocessing import TransactionEncoder
import pandas as pd
from mlxtend.frequent_patterns import apriori, association_rules
from itertools import permutations
pd.options.display.float_format = '{:.5f}'.format
pd.options.display.max_columns = None

data = pd.read_csv('teleco_market_basket.csv')

### Part I

### A. 1.

What items drive sales of other items? Answering this question would allow the company to better plan promotional deals and improve the suggested products.

### A. 2.

The main goal of this analysis is to find groupings of items that sell well together to help management make better decisions.

### Part II

### B. 1.

Market Basket Analysis uses data sets to try and find associations between products. The Apriori algorithm is normally used in Market Basket Analysis to add in the process.   

Once a data set is cleaned and prepared for the analysis, using the Apriori algorithm will return a set of rules with a large amount of information about each rule. A rule can be simply explained by an example, Apple -> Banana. The rule would read if someone buys an apple, they are likely to buy a banana as well. The algorithm also computes confidence, support, and lift for each rule that is returned.

### B. 2.

The block below shows one transaction from the data set.

In [15]:
data.iloc[1]

Item01               Logitech M510 Wireless mouse
Item02                                  HP 63 Ink
Item03                                  HP 65 ink
Item04                 nonda USB C to USB Adapter
Item05                  10ft iPHone Charger Cable
Item06                               HP 902XL ink
Item07               Creative Pebble 2.0 Speakers
Item08        Cleaning Gel Universal Dust Cleaner
Item09              Micro Center 32GB Memory card
Item10    YUNSONG 3pack 6ft Nylon Lightning Cable
Item11               TopMate C5 Laptop Cooler pad
Item12                  Apple USB-C Charger cable
Item13               HyperX Cloud Stinger Headset
Item14                TONOR USB Gaming Microphone
Item15             Dust-Off Compressed Gas 2 pack
Item16             3A USB Type C Cable 3 pack 6FT
Item17                      HOVAMP iPhone charger
Item18                   SanDisk Ultra 128GB card
Item19       FEEL2NICE 5 pack 10ft Lighning cable
Item20        FEIYOLD Blue light Blocking Glasses


### B. 3

Market Basket Analysis doesn't hold the normal assumptions most algorithms have. The data doesn't have to have a linear relationship and normality isn't a concern. One of the assumptions that can be found through working a Market Basket problem is that the data once cleaned must be binary in nature. This allows the algorithm see if something was either purchased or it wasn't.

### Part III

### C. 1.

This block takes the original dataframe, converts it into a list of lists, then uses the Transaction encoder from mlxtend to encode the data. That data is then turned back into a dataframe that the Apriori algorithm can use. In this block we are also saving our cleaned data set for submission as well.

In [16]:
#converting dataframe into list of lists
l=[]
for i in range(0,len(data)):
    l.append([str(data.values[i,j]) for j in range(0,20)])
    
flat = [item for l in l for item in l]
items = list(set(flat))

encoder = TransactionEncoder().fit(l)
onehot = encoder.transform(l)
onehot = pd.DataFrame(onehot, columns=encoder.columns_).drop('nan', axis=1)

onehot.to_csv('Task 3 Cleaned Set.csv')

### C. 2.

The code below uses the Apriori algorithm on the cleaned and encoded data set. It then generates our association rules and sorts them.

In [17]:
frq = apriori(onehot,min_support=0.02, use_colnames=True)
rules = association_rules(frq, metric='lift')
rules.sort_values('confidence', ascending = False, inplace = True)

### C. 3.

The code block below prints the association rule table with support, confidence, and lift displayed.

In [18]:
rules

antecedents                         consequents  \
5           (SanDisk Ultra 64GB card)    (Dust-Off Compressed Gas 2 pack)   
7     (Screen Mom Screen Cleaner kit)    (Dust-Off Compressed Gas 2 pack)   
9  (VIVO Dual LCD Monitor Desk mount)    (Dust-Off Compressed Gas 2 pack)   
3                         (HP 61 ink)    (Dust-Off Compressed Gas 2 pack)   
1                      (Apple Pencil)    (Dust-Off Compressed Gas 2 pack)   
8    (Dust-Off Compressed Gas 2 pack)  (VIVO Dual LCD Monitor Desk mount)   
2    (Dust-Off Compressed Gas 2 pack)                         (HP 61 ink)   
0    (Dust-Off Compressed Gas 2 pack)                      (Apple Pencil)   
6    (Dust-Off Compressed Gas 2 pack)     (Screen Mom Screen Cleaner kit)   
4    (Dust-Off Compressed Gas 2 pack)           (SanDisk Ultra 64GB card)   

   antecedent support  consequent support  support  confidence    lift  \
5             0.04913             0.11918  0.02046     0.41655 3.49504   
7             0.06479             0.11918  0.02400     0.37037 3.10755   
9             0.08706             0.11918  0.02986     0.34303 2.87817   
3             0.08192             0.11918  0.02633     0.32140 2.69666   
1             0.08985             0.11918  0.02546     0.28338 2.37769   
8             0.11918             0.08706  0.02986     0.25056 2.87817   
2             0.11918             0.08192  0.02633     0.22092 2.69666   
0             0.11918             0.08985  0.02546     0.21365 2.37769   
6             0.11918             0.06479  0.02400     0.20134 3.10755   
4             0.11918             0.04913  0.02046     0.17170 3.49504   

   leverage  conviction  
5   0.01461     1.50968  
7   0.01627     1.39894  
9   0.01949     1.34073  
3   0.01657     1.29799  
1   0.01475     1.22913  
8   0.01949     1.21817  
2   0.01657     1.17841  
0   0.01475     1.15743  
6   0.01627     1.17098  
4   0.01461     1.14798

### C. 4.

In [19]:
rules[:3]

antecedents                       consequents  \
5           (SanDisk Ultra 64GB card)  (Dust-Off Compressed Gas 2 pack)   
7     (Screen Mom Screen Cleaner kit)  (Dust-Off Compressed Gas 2 pack)   
9  (VIVO Dual LCD Monitor Desk mount)  (Dust-Off Compressed Gas 2 pack)   

   antecedent support  consequent support  support  confidence    lift  \
5             0.04913             0.11918  0.02046     0.41655 3.49504   
7             0.06479             0.11918  0.02400     0.37037 3.10755   
9             0.08706             0.11918  0.02986     0.34303 2.87817   

   leverage  conviction  
5   0.01461     1.50968  
7   0.01627     1.39894  
9   0.01949     1.34073

### Part IV

### D. 1.

The three metrics used to gauge association rules are support, confidence, and lift. When looking at the top 3 rules above as sorted by the highest confidence, we can also see that these three have high lift values as well. The support for these rules are on the lower end of the spectrum.

Support can be summarized as the percentage of transactions that will have both the antecedent and consequesnt in them. As we can see with the rules above, the highest support value is 0.02986 or in other words, both the entire itemset appears together in about 3% of transactions.

Confidence can be summarized as the probability of a transaction containing the antecedent as well as the consequent. For example with our top rule, it can be said that about 42% of transactions that contain SanDisk Ultra 64GB card also have Dust-Off Compressed Gas 2 pack in them.

Lift can be summarized as the ratio of confidence to expected confidence. A lift of less than 1 means that there is a negative correlation between items, a lift greater than one shows that a stronger association exists. For our example with the top 3 rules, the top rule also has the highest lift. That lift is 3.495 which is much greater than 1 meaning a strong association exists between our two items. (solver, 2016)

### D. 2.

We can see from the rules that associations between products do exist. Quite a few of those rules have high confidence and lift values, which tend to show that the they are frequently bought together. The analysis goes on to show the company that bundles could be formed to offer greater savings to customers and potentially more sales for the company. While many rules have a lower support, even rules only occuring in 2% of all transactions can be viewed as popular when you have large amounts of transaction.

### D. 3.

The company should spend some time looking at rules and the products they include, to see where the greatest profit margains exist. This would allow them to be more flexible on the price and potential increase sales. These rules could also be used to generate "Frequently Bought Together" displays that could drive interest in those products. 

### Part V

### G.

### References

Association rules. solver. (2016, January 14). Retrieved October 23, 2021, from https://www.solver.com/xlminer/help/association-rules.  

McColl, L. (2018, May 31). Market basket analysis: Understanding customer behaviour. Select Statistical Consultants. Retrieved October 24, 2021, from https://select-statistics.co.uk/blog/market-basket-analysis-understanding-customer-behaviour/. 